library/core/src/option.rs

```diff
@@ -1804,10 +1804,21 @@ impl<T> Option<T> {
     pub const fn get_or_insert_with<F>(&mut self, f: F) -> &mut T
     where
         F: [const] FnOnce() -> T + [const] Destruct,
-        T: [const] Destruct,
     {
         if let None = self {
-            *self = Some(f());
+            // The effect of this is identical to
+            //     *self = Some(f());
+            // except that it does not drop the old value of `*self`. This is not a leak, because
+            // we just checked that the old value is `None`, which contains no fields to drop.
+            //
+            // This implementation avoids needing a `T: [const] Destruct` bound and avoids
+            // possibly compiling needless drop code. Ideally, the compiler would prove that that
+            // drop is not necessary itself, but it currently doesn't, even with
+            // `const_precise_live_drops` enabled.
+            //
+            // It could also be expressed as `core::ptr::write(self, Some(f()))`; that would be
+            // additional unsafe code, so should be avoided unless a reason is found.
```

Comment:
maybe include the unsafe block inside the code example?

Reply:
Done.